src/analysis/processing/qgsalgorithmfilter.cpp

```diff
@@ -0,0 +1,247 @@
+  addOutputButton->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/mActionAddLayer.svg" ) ) );
+  addOutputButton->setText( tr( "Add Output" ) );
+
+  QToolButton *removeOutputButton = new QToolButton();
+  removeOutputButton->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/mActionRemoveLayer.svg" ) ) );
+  removeOutputButton->setToolTip( tr( "Remove Selected Outputs" ) );
+
+  layout->addWidget( addOutputButton, 2, 1, 1, 1 );
+  layout->addWidget( removeOutputButton, 3, 1, 1, 1 );
+
+  connect( addOutputButton, &QToolButton::clicked, this, &QgsFilterAlgorithmConfigurationWidget::addOutput );
+  connect( removeOutputButton, &QToolButton::clicked, this, &QgsFilterAlgorithmConfigurationWidget::removeSelectedOutputs );
+
+  connect( mOutputExpressionWidget->selectionModel(), &QItemSelectionModel::selectionChanged, [removeOutputButton, this]
+  {
+    removeOutputButton->setEnabled( !mOutputExpressionWidget->selectionModel()->selectedIndexes().isEmpty() );
+  } );
+}
+
+QVariantMap QgsFilterAlgorithmConfigurationWidget::configuration() const
+{
+  QVariantList outputs;
+
+  for ( int i = 0; i < mOutputExpressionWidget->rowCount(); ++i )
+  {
+    QVariantMap output;
+    output.insert( QStringLiteral( "name" ), mOutputExpressionWidget->item( i, 0 )->text() );
+    output.insert( QStringLiteral( "expression" ), mOutputExpressionWidget->item( i, 1 )->text() );
+    outputs.append( output );
+  }
+
+  QVariantMap map;
+  map.insert( "outputs", outputs );
+
+  return map;
+}
+
+
+QVariantMap QgsFilterAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  std::unique_ptr< QgsFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !source )
+    return QVariantMap();
+
+  QgsExpressionContext expressionContext = context.expressionContext();
+  for ( Output *output : qgis::as_const( mOutputs ) )
+  {
+    output->sink.reset( parameterAsSink( parameters, output->name, context, output->destinationIdentifier, source->fields(), source->wkbType(), source->sourceCrs() ) );
+    output->expression.prepare( &expressionContext );
+  }
+
+  long count = source->featureCount();
+
+  QgsFeature f;
+  QgsFeatureIterator it = source->getFeatures();
+
+  double step = count > 0 ? 100.0 / count : 1;
+  int current = 0;
+
+  while ( it.nextFeature( f ) )
+  {
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+
+    expressionContext.setFeature( f );
+
+    for ( Output *output : qgis::as_const( mOutputs ) )
+    {
+      if ( output->expression.evaluate( &expressionContext ).toBool() )
+        output->sink->addFeature( f, QgsFeatureSink::FastInsert );
+    }
+
+    feedback->setProgress( current * step );
+    current++;
+  }
+
+  QVariantMap outputs;
+  for ( const Output *output : qgis::as_const( mOutputs ) )
+  {
+    outputs.insert( output->name, output->destinationIdentifier );
+  }
```

Comment:
I think you should be deleting the sinks here too

Reply:
They are deleted in the destructor, but that can also be moved here.

Reply:
I think it needs to be moved - some sinks do finalization code which probably should be done be within the processAlgorithm step

Reply:
also done